---
title: "Flacc m6a - Differr + Yanocomp"
---
```{r}
library(dplyr)
library(ggplot2)
library(data.table)
library(clusterProfiler)
library(org.At.tair.db)
library(readr)
library(GenomicFeatures)
library(purrr)
library(reshape2)
library(Biostrings)
library(eulerr)
```
Load genome and annotation
```{r}
At.genome <- Rsamtools::FaFile("/Volumes/cluster/ggs_lab/cmetheringham001/Shared_Ref/TAIR10_chr_all.fas")
# Load TxDb
txdb <- makeTxDbFromGFF("/Volumes/cluster/ggs_lab/cmetheringham001/Shared_Ref/Araport11_GFF3_genes_transposons.201606.no_chr.gtf")
```

Colour palette
```{r colour bar, echo=FALSE, fig.height=2, fig.width=13}
palette <- c(
 # Writers
  "MTA"     = "#004488",
  "VIR"  = "#009988",
  "FIP37"   = "#0077BB",
  "hiz2-4" = "#AA4499",
  "hiz2-2"    = "#CCAA22",
  "HAKAI" = "#556B2F",
  #Readers
  "ECT"     = "#EE7733",
  "CPSF30"  = "#DD3377",
  #Other
  "U2B"     = "#CC3311",
  "Col-0"    = "#888888"
)
barplot(rep(1, 10), col = palette, names.arg = names(palette), yaxt = "n")
```
Get the diff err stats for hiz2-4 vs col0
olumns of bed file are:

    chrom, start, end, name
    score: -log10 of the FDR, rounded to nearest whole number
    strand
    odds ratio: the change in the ratio of matches to mismatches in the wild type compared to the mutant with low modifications. An odds ratio > 0 indicates more modifications in the WT.
    G statistic for the comparison of pooled WT and mutant samples.
    -log10 P value for the comparison of pooled WT and mutant samples.
    -log10 FDR for the comparison of pooled WT and mutant samples.
    G statistic for the homogeneity test of mutant replicates.
    -log10 P value for the homogeneity test of mutant replicates.
    G statistic for the homogeneity test of wild type replicates.
    -log10 P value for the homogeneity test of wild type replicates.

```{r}
hiz2_4_diff_err <- fread("/Volumes/cluster/ggs_lab/mtparker/papers/flacc/flacc_nanopore/pipeline/differr/flacc_vs_col0.bed")
colnames(hiz2_4_diff_err) <- c("chrom", "start", "end", "name", "score", "strand", "odds", "G", "P", "FDR", "g_hom", "P_hom", "g_hom_c", "P_hom_c")
head(hiz2_4_diff_err)
```

Where are the sites located?
```{r}
# List of positions
positions <- tibble(
  hiz2_4_diff_err[,1:6]
)
positions <- positions %>%
  mutate(start = as.numeric(start))
head(positions)
```


```{r}
# Extract UTRs
five_utrs <- fiveUTRsByTranscript(txdb, use.names=TRUE)
three_utrs <- threeUTRsByTranscript(txdb, use.names=TRUE)

# Flatten into GRanges
five_utrs_flat <- unlist(five_utrs, use.names=FALSE)
three_utrs_flat <- unlist(three_utrs, use.names=FALSE)

five_utrs_flat$region_type <- "5UTR"
three_utrs_flat$region_type <- "3UTR"

# Combine 5' and 3' UTRs into one GRanges
utr_regions <- c(five_utrs_flat, three_utrs_flat)
```


```{r}
# Make GRanges from your table
data_gr <- GRanges(
  seqnames = positions$chrom,
  ranges = IRanges(start = positions$start, end = positions$end),
  strand = positions$strand,
  transcript_id = positions$name
)

# Find overlaps
hits <- findOverlaps(data_gr, utr_regions, ignore.strand=FALSE)

# Create a vector of UTR annotations
utr_annotation <- rep(NA_character_, length(data_gr))
utr_annotation[queryHits(hits)] <- mcols(utr_regions)$region_type[subjectHits(hits)]

# Add to your table
positions_annotated <- positions %>%
  mutate(UTR_type = utr_annotation)

head(positions_annotated)
```
```{r}
table(positions_annotated$UTR_type)
```

Distance from nearest vir
Load in the vir differr data
```{r}
vir1_diff_err <- fread("/Volumes/cluster/ggs_lab/mtparker/papers/flacc/flacc_nanopore/pipeline/differr/vir1_vs_VIRc.bed")
colnames(vir1_diff_err) <- c("chrom", "start", "end", "name", "score", "strand", "odds", "G", "P", "FDR", "g_hom", "P_hom", "g_hom_c", "P_hom_c")
head(vir1_diff_err)
```

```{r}
gr1 <- GRanges(
  seqnames = hiz2_4_diff_err$chrom,
  ranges = IRanges(start = hiz2_4_diff_err$start, end = hiz2_4_diff_err$end),
  strand = hiz2_4_diff_err$strand
)

gr2 <- GRanges(
  seqnames = vir1_diff_err$chrom,
  ranges = IRanges(start = vir1_diff_err$start, end = vir1_diff_err$end),
  strand = vir1_diff_err$strand
)
hits <- findOverlaps(gr1, gr2, maxgap = 5, ignore.strand = FALSE)
n_hits <- length(unique(queryHits(hits)))
total <- length(gr1)
n_hits/total
```
```{r}
gr_to_set <- function(gr) {
  paste0(seqnames(gr), ":", start(gr), "-", end(gr), ":", strand(gr))
}
hiz2_set <- gr_to_set(gr1)
vir1_set <- gr_to_set(gr2)
```


```{r}
venn_list <- euler(list(
  hiz2 = hiz2_set,
  vir = vir1_set
))

venn_plot <- plot(
  venn_list,
  fills = list(fill = c("#004488", "#009988"), alpha = 0.7),
  labels = list(font = 2, cex = 1.2),
  edges = TRUE,
  quantities = TRUE
)

venn_plot
```
Overlaps within 5nt in each sample
```{r}
A <- length(gr1)                # total in hiz2
B <- length(gr2)                # total in vir1
AB <- length(unique(queryHits(findOverlaps(gr1, gr2, maxgap = 5, ignore.strand = FALSE))))  # hiz2 hits vir1

# Euler input with counts
venn_input <- c(
  hiz2 = A - AB,      
  vir  = B,             
  "hiz2&vir" = AB       
)

venn_input["vir"] <- B - AB
venn_obj <- euler(venn_input)

venn_5nt <- plot(
  venn_obj,
  fills = list(fill = c("#AA4499", "#009988"), alpha = 0.7),
  labels = FALSE,
  edges = TRUE,
  quantities = TRUE
)

venn_5nt
```
```{r}
ggsave(
  filename = "../../Figures/figure_2/hiz2-4_vir1_diff_err_overlap.pdf",
  plot = venn_5nt,
  width = 6, height = 5, units = "in" 
)
ggsave(
  filename = "../../Figures/figure_2/hiz2-4_vir1_diff_err_overlap.png",
  plot = venn_5nt,
  width = 6, height = 5, units = "in" 
)
```



Underlying motifs

Expression of genes
What are the genes
