---
title: "Flacc alignment diagram"
---
Colour palette
```{r colour bar, echo=FALSE, fig.height=2, fig.width=13}
palette <- c(
 # Writers
  "MTA"     = "#004488",
  "VIR"  = "#009988",
  "FIP37"   = "#0077BB",
  "hiz2-4" = "#AA4499",
  "hiz2-2"    = "#CCAA22",
  "HAKAI" = "#556B2F",
  #Readers
  "ECT"     = "#EE7733",
  "CPSF30"  = "#DD3377",
  #Other
  "U2B"     = "#CC3311",
  "Col0"    = "#888888"
)
barplot(rep(1, 10), col = palette, names.arg = names(palette), yaxt = "n")
```

```{r}
library(Gviz)
library(Rsamtools)
library(GenomicAlignments)
library(scales)
```
Add gene model plot
```{r}
exons <- data.frame(start = c(21683144, 21684539, 21684724, 21686222, 21687448),
                    end = c(21684443, 21684625, 21686014, 21687101, 21688273))

utr3 <- data.frame(start = c(21683144, 21687971), end = c(21683678, 21688273))
```


Old Flacc data
```{r}
bam_files <- c("flacc_1a.genome.bam", "flacc_2.genome.bam", "flacc_3.genome.bam", "flacc_4.genome.bam",
               "col0_1.genome.bam", "col0_2.genome.bam", "col0_3.genome.bam")
bam_paths <- file.path("../../bam_files", bam_files)
names(bam_paths) <- bam_files

chr <- "5"   # renamed hr -> chr for consistency
from <- 21683144
to <- 21688273
region <- GRanges(chr, IRanges(from, to))

# Setup plot
total_height <- 0
read_blocks <- list()
colors <- c("#AA4499","#AA4499","#AA4499","#AA4499","#888888","#888888","#888888")

for (i in seq_along(bam_paths)) {
  reads <- readGAlignments(bam_paths[i], param = ScanBamParam(which = region))
  cigarStrings <- cigar(reads)
  softclipped <- grepl("S", cigarStrings)
  read_blocks[[i]] <- list(
    reads = reads,
    softclipped = softclipped,
    y_start = total_height + 1,
    y_end = total_height + length(reads)
  )
  total_height <- total_height + length(reads) + 2
}

# Begin plot with no y axis ticks or labels
plot(NA, xlim = c(from, to), ylim = c(0, total_height + 10),
     xlab = "Genomic Position", ylab = "",
     main = "Alignment of reads to HIZ2",
     yaxt = "n",  # suppress y axis
     cex.main = 1.3, cex.lab = 1.1)

# Add light vertical grid lines every ~1000 bases
#abline(v = seq(from, to, by = 1000), col = "#888888", lty = "dotted")

for (i in seq_along(read_blocks)) {
  rb <- read_blocks[[i]]
  reads <- rb$reads
  y_offset <- rb$y_start - 1
  col_normal <- colors[(i - 1) %% length(colors) + 1]

  for (j in seq_along(reads)) {
    segments(start(reads[j]), y_offset + j, end(reads[j]), y_offset + j, col = col_normal, lwd = 1.5)
  }
  
  # Sample name on left, aligned nicely with some margin space
  text(from - 500, y_offset + length(reads) / 2, labels = names(bam_paths)[i], pos = 2, cex = 0.9)
}

# Add T-DNA insertion marker
insertion_site <- 21683920
abline(v = insertion_site, col = "darkred", lty = 2, lwd = 2)
text(insertion_site, total_height + 4, "T-DNA", col = "darkred", pos = 2)
```
```{r}
plot(NA, xlim = c(to, from), ylim = c(0, total_height + 10),
     xlab = "Genomic Position", ylab = "",
     main = "Alignment of reads to HIZ2",
     yaxt = "n",  # suppress y axis
     cex.main = 1.3, cex.lab = 1.1)

# Sample plotting code unchanged:
for (i in seq_along(read_blocks)) {
  rb <- read_blocks[[i]]
  reads <- rb$reads
  y_offset <- rb$y_start - 1
  col_normal <- colors[(i - 1) %% length(colors) + 1]

# Filter to keep only reads NOT on the positive strand
  keep_idx <- which(as.character(strand(reads)) == "-")
  reads <- reads[keep_idx]
  
  # Recalculate exon blocks only for filtered reads
  exon_blocks <- cigarRangesAlongReferenceSpace(cigar(reads), pos = start(reads),
                                               ops = "M", with.ops = FALSE)

for (j in seq_along(reads)) {
  y <- y_offset + j
  blocks <- exon_blocks[[j]]
  for (k in seq_along(blocks)) {
    segments(start(blocks[k]), y, end(blocks[k]), y, col = col_normal, lwd = 1.5)
  }
}
  
  # Sample name on left, aligned nicely with some margin space
  #text(from - 500, y_offset + length(reads) / 2, labels = names(bam_paths)[i], pos = 2, cex = 0.9)
}

# T-DNA insertion markers:
abline(v = c(21683920), col = "darkred", lty = 2, lwd = 1)
text(21683920, total_height + 4, "T-DNA", col = "darkred", pos = 2)
```
```{r}
layout(matrix(1:2, nrow = 2), heights = c(5, 2))
# Set outer margins
par(oma = c(1, 1, 1, 1))  # outer margins: bottom, left, top, right
par(mar = c(0.5, 1.5, 1, 0.5))  # inner margins: bottom, left, top, right

# ---- Top plot: your read alignments ----
plot(NA, xlim = c(to, from), ylim = c(0, total_height + 10),
     xlab = "", ylab = "",
     main = "Alignment of reads to HIZ2",
     yaxt = "n", xaxt = "n",  
     cex.main = 1.3, cex.lab = 1.1)

# Draw alignments as before
for (i in seq_along(read_blocks)) {
  rb <- read_blocks[[i]]
  reads <- rb$reads
  y_offset <- rb$y_start - 1
  col_normal <- colors[(i - 1) %% length(colors) + 1]

# Filter to keep only reads NOT on the positive strand
  keep_idx <- which(as.character(strand(reads)) == "-")
  reads <- reads[keep_idx]
  
  # Recalculate exon blocks only for filtered reads
  exon_blocks <- cigarRangesAlongReferenceSpace(cigar(reads), pos = start(reads),
                                               ops = "M", with.ops = FALSE)

for (j in seq_along(reads)) {
  y <- y_offset + j
  blocks <- exon_blocks[[j]]
  for (k in seq_along(blocks)) {
    segments(start(blocks[k]), y, end(blocks[k]), y, col = col_normal, lwd = 1.5)
  }
}
  #text(from - 500, y_offset + length(reads) / 2, labels = names(bam_paths)[i], pos = 2, cex = 0.9)
}

abline(v = c(21683920), col = "darkred", lty = 2, lwd = 1)
text(21683920, total_height + 4, "T-DNA", col = "darkred", pos = 2)

# ---- Bottom plot: gene model schematic ----
par(mar = c(2.5, 1.5, 0.5, 0.5))
plot(NA, xlim = c(to, from), ylim = c(0, 1), xaxt = "n", yaxt = "n", xlab = "Genomic Position (kb)", ylab = "", bty = "n")

# Draw exons as filled rectangles
for (i in seq_len(nrow(exons))) {
  rect(exons$start[i], 0.3, exons$end[i], 0.7, col = "black", border = NA)
}

# Draw 3' UTR as a different colored rectangle
rect(utr3$start, 0.3, utr3$end, 0.7, col = "#666666", border = NA)

# Draw introns as lines connecting exons
for (i in seq_len(nrow(exons)-1)) {
  segments(exons$end[i], 0.5, exons$start[i+1], 0.5, col = "black", lwd = 2)
}

# Optional: add legend or text labels
text(x = mean(c(from, to)), y = 0.9, labels = "", cex = 0.8)
```

New Flacc data
```{r}
bam_files <- c("flacc_mutant_1.aligned.genome.bam", "flacc_mutant_2.aligned.genome.bam",
               "flacc_mutant_3.aligned.genome.bam", "flacc_mutant_4.aligned.genome.bam",
               "flacc_OX_1.aligned.genome.bam", "flacc_OX_2.aligned.genome.bam",
               "flacc_OX_3.aligned.genome.bam", "flacc_OX_4.aligned.genome.bam",
               "col0_1.aligned.genome.bam", "col0_2.aligned.genome.bam",
               "col0_3.aligned.genome.bam", "col0_4.aligned.genome.bam")
bam_paths <- file.path("/Volumes/cluster/ggs_lab/cmetheringham001/SIKFLACC/sikflacc_nanopore/data/modkit", bam_files)
names(bam_paths) <- bam_files

chr <- "5"   # renamed hr -> chr for consistency
from <- 21683144
to <- 21688273
region <- GRanges(chr, IRanges(from, to))

# Setup plot
total_height <- 0
read_blocks <- list()
colors <- c("#CCAA22", "#CCAA22", "#CCAA22","#CCAA22",
            "#AA4499","#AA4499","#AA4499","#AA4499",
            "#888888", "#888888","#888888","#888888")

for (i in seq_along(bam_paths)) {
  reads <- readGAlignments(bam_paths[i], param = ScanBamParam(which = region))
  cigarStrings <- cigar(reads)
  softclipped <- grepl("S", cigarStrings)
  read_blocks[[i]] <- list(
    reads = reads,
    softclipped = softclipped,
    y_start = total_height + 1,
    y_end = total_height + length(reads)
  )
  total_height <- total_height + length(reads) + 2
}

plot(NA, xlim = c(from, to), ylim = c(0, total_height + 10),
     xlab = "Genomic Position", ylab = "",
     main = "Alignment of reads to HIZ2",
     yaxt = "n",  # suppress y axis
     cex.main = 1.3, cex.lab = 1.1)

# Add light vertical grid lines every ~1000 bases
#abline(v = seq(from, to, by = 1000), col = "#888888", lty = "dotted")

for (i in seq_along(read_blocks)) {
  rb <- read_blocks[[i]]
  reads <- rb$reads
  y_offset <- rb$y_start - 1
  col_normal <- colors[(i - 1) %% length(colors) + 1]

# Filter to keep only reads NOT on the positive strand
  keep_idx <- which(as.character(strand(reads)) == "-")
  reads <- reads[keep_idx]
  
  # Recalculate exon blocks only for filtered reads
  exon_blocks <- cigarRangesAlongReferenceSpace(cigar(reads), pos = start(reads),
                                               ops = "M", with.ops = FALSE)

for (j in seq_along(reads)) {
  y <- y_offset + j
  blocks <- exon_blocks[[j]]
  for (k in seq_along(blocks)) {
    segments(start(blocks[k]), y, end(blocks[k]), y, col = col_normal, lwd = 1.5)
  }
}
  
  # Sample name on left, aligned nicely with some margin space
  text(from - 500, y_offset + length(reads) / 2, labels = names(bam_paths)[i], pos = 2, cex = 0.9)
}

# Add T-DNA insertion marker
insertion_site <- 21683920
abline(v = insertion_site, col = "darkred", lty = 2, lwd = 1)
text(insertion_site, total_height + 4, "T-DNA", col = "darkred", pos = 2)
insertion_site <- 21687450
abline(v = insertion_site, col = "darkred", lty = 2, lwd = 1)
text(insertion_site, total_height + 4, "T-DNA", col = "darkred", pos = 2)
```
```{r}
plot(NA, xlim = c(to, from), ylim = c(0, total_height + 10),
     xlab = "Genomic Position", ylab = "",
     main = "Alignment of reads to HIZ2",
     yaxt = "n",  # suppress y axis
     cex.main = 1.3, cex.lab = 1.1)

# Sample plotting code unchanged:
for (i in seq_along(read_blocks)) {
  rb <- read_blocks[[i]]
  reads <- rb$reads
  y_offset <- rb$y_start - 1
  col_normal <- colors[(i - 1) %% length(colors) + 1]

  for (j in seq_along(reads)) {
    segments(start(reads[j]), y_offset + j, end(reads[j]), y_offset + j, col = col_normal, lwd = 1.5)
  }
  
  # Sample name on left, aligned nicely with some margin space
  #text(from - 500, y_offset + length(reads) / 2, labels = names(bam_paths)[i], pos = 2, cex = 0.9)
}

# T-DNA insertion markers:
abline(v = c(21683920, 21687450), col = "darkred", lty = 2, lwd = 1)
text(21683920, total_height + 4, "T-DNA", col = "darkred", pos = 2)
text(21687450, total_height + 4, "T-DNA", col = "darkred", pos = 2)
```

```{r}
pdf("../../Figures/figure_2/new_flacc_read_plot.pdf", width = 8, height = 6)  # adjust size as needed

layout(matrix(1:2, nrow = 2), heights = c(5, 2))
# Set outer margins
par(oma = c(1, 1, 1, 1))  # outer margins: bottom, left, top, right
par(mar = c(0.5, 1.5, 1, 0.5))  # inner margins: bottom, left, top, right

# ---- Top plot: your read alignments ----
plot(NA, xlim = c(to, from), ylim = c(0, total_height + 10),
     xlab = "", ylab = "",
     main = "Alignment of reads to HIZ2",
     yaxt = "n", xaxt = "n",  
     cex.main = 1.3, cex.lab = 1.1)

# Draw alignments as before
for (i in seq_along(read_blocks)) {
  rb <- read_blocks[[i]]
  reads <- rb$reads
  y_offset <- rb$y_start - 1
  col_normal <- colors[(i - 1) %% length(colors) + 1]

# Filter to keep only reads NOT on the positive strand
  keep_idx <- which(as.character(strand(reads)) == "-")
  reads <- reads[keep_idx]
  
  # Recalculate exon blocks only for filtered reads
  exon_blocks <- cigarRangesAlongReferenceSpace(cigar(reads), pos = start(reads),
                                               ops = "M", with.ops = FALSE)

for (j in seq_along(reads)) {
  y <- y_offset + j
  blocks <- exon_blocks[[j]]
  for (k in seq_along(blocks)) {
    segments(start(blocks[k]), y, end(blocks[k]), y, col = col_normal, lwd = 1.5)
  }
}
  #text(from - 500, y_offset + length(reads) / 2, labels = names(bam_paths)[i], pos = 2, cex = 0.9)
}

abline(v = c(21683920, 21687450), col = "darkred", lty = 2, lwd = 1)
text(21683920, total_height + 4, "T-DNA", col = "darkred", pos = 2)
text(21687450, total_height + 4, "T-DNA", col = "darkred", pos = 2)

# ---- Bottom plot: gene model schematic ----
par(mar = c(2.5, 1.5, 0.5, 0.5))
plot(NA, xlim = c(to, from), ylim = c(0, 1), xaxt = "n", yaxt = "n", xlab = "Genomic Position (kb)", ylab = "", bty = "n")

# Draw exons as filled rectangles
for (i in seq_len(nrow(exons))) {
  rect(exons$start[i], 0.3, exons$end[i], 0.7, col = "black", border = NA)
}

# Draw 3' UTR as a different colored rectangle
rect(utr3$start, 0.3, utr3$end, 0.7, col = "#666666", border = NA)

# Draw introns as lines connecting exons
for (i in seq_len(nrow(exons)-1)) {
  segments(exons$end[i], 0.5, exons$start[i+1], 0.5, col = "black", lwd = 2)
}

# Optional: add legend or text labels
text(x = mean(c(from, to)), y = 0.9, labels = "", cex = 0.8)

# Automatically determine suitable tick positions
ticks <- pretty(c(to, from), n = 5)  # adjust 'n' to control how many ticks

# Format tick labels in Mb, rounded to 2 decimal places
labels <- format(round(ticks / 1e6, 2), nsmall = 2)

# Add x-axis with formatted labels
axis(1, at = ticks, labels = labels, las = 1, cex.axis = 0.8)

dev.off()
```
