---
title: "PolyA_site_finder"
---
Load libraries required to mix R and bash
```{r}
library(glue)
library(data.table)
library(purrr)
library(fs)
library(stringr)
library(parallel)
```

Set the input bam files and output bed
```{r}
bam_file <- "/Volumes/cluster/ggs_lab/mtparker/papers/flacc/flacc_nanopore/pipeline/aligned_data/col0_1.genome.bam"
bam_list <- dir_ls("/Volumes/cluster/ggs_lab/mtparker/papers/flacc/flacc_nanopore/pipeline/aligned_data", glob = "*genome.filtered.bam")  %>%
  str_subset("col0")
bed_file <- "bed/genome_polya_sites.bed"
polya_bed_file <- "bed/polya_site_counts.bed"
results_file <- "/Volumes/cluster/ggs_lab/mtparker/papers/flacc/flacc_nanopore/pipeline/apa_results/vir1_vs_col0.apa_results.bed"
results_header <- c("chrom","start","end","gene_id","score","strand","wass_dist","wass_dir","wass_pval","wass_fdr","nreads_cntrl","nreads_treat")
region_file <- "bed/regions_for_samtools.bed"
piled_up_results <- "bed/piled_results.tsv"
```
Define the transcripts that have a change in poly(A) site usage
```{r}
apa_results <- read.delim(results_file)
colnames(apa_results) <- results_header
#filter by significance
apa_results <- apa_results[apa_results$wass_fdr < 0.001,]
regions <- paste0(apa_results$chrom, "\t", apa_results$start -250, "\t", apa_results$end +250)
writeLines(regions, region_file)
```

Use samtools to extract the 3' ends to bed file
```{R}
cmd <- glue('samtools view -F 0x4 -L "{region_file}" "{bam_file}" | \\
gawk \'BEGIN{{OFS="\\t"}} {{
    read_name = $1;
    flag = $2;
    chrom = $3;
    start = $4;
    cigar = $6;
    strand = ((and(flag, 16)) ? "-" : "+");
    len = 0;
    while (match(cigar, /[0-9]+[MDN=X]/)) {{
        len += substr(cigar, RSTART, RLENGTH - 1);
        cigar = substr(cigar, RSTART + RLENGTH);
    }}
    end_pos = start + len - 1;
    polya_site = (strand == "+" ? end_pos : start);
    print chrom, polya_site, polya_site + 1, read_name, 0, strand;
}}\' > "{bed_file}"')
system(cmd)
```
```{r}
# Build the shell command
sort_groupby_cmd <- glue::glue('sort -k1,1 -k2,2n "{bed_file}" | \\
  bedtools groupby -g 1,2,3,6 -c 4 -o count > "{polya_bed_file}"')
system(sort_groupby_cmd)

polya_counts <- fread(polya_bed_file, col.names = c("chrom", "start", "end", "strand", "read_count"))

# Display first few rows
head(polya_counts)
```



```{r}
# Group and count
#polya_counts <- all_bed[, .(count = .N), by = .(chrom, start, end, strand)]

# Save to BED
fwrite(polya_counts, polya_bed_file, sep = "\t", col.names = FALSE)

#polya_counts <- fread(polya_bed_file, col.names = c("chrom", "start", "end", "strand", "read_count"))
head(polya_counts)
```


Use a 5bp sliding window to aggregate the a tail ends
```{r}
#Set the sorting order
setorder(polya_counts, chrom, strand, start)

test_subset <- polya_counts[1:500]

# Sliding window function
slide_window_counts <- function(df, window_size = 5, step = 5) {
  starts <- seq(min(df$start), max(df$start), by = step)
  result <- data.table()
  for (s in starts) {
    e <- s + window_size
    count_sum <- df[start >= s & start < e, sum(read_count)]
    if (!is.na(count_sum)) {
      result <- rbind(result, data.table(win_start = s, win_end = e, count_sum = count_sum))
    }
  }
  return(result)
}

# Apply by chrom and strand
sliding_results <- test_subset[, slide_window_counts(.SD), by = .(chrom, strand)]
sliding_results <- sliding_results[sliding_results$count_sum>5,]
head(sliding_results)
```
```{r}
library(data.table)

# Ensure data is sorted
setorder(polya_counts, chrom, strand, start)

# Efficient window count function
slide_window_dt <- function(dt, window_size = 1, step = 1) {
  # Determine window start and end positions for each chrom + strand
  dt[, {
    min_start <- min(start)
    max_start <- max(start)
    bins <- seq(min_start, max_start, by = step)
    
    # Assign each row to a bin
    bin_id <- findInterval(start, bins)

    # Compute sum within each bin
    out <- .SD[, .(win_start = bins[unique(bin_id)],
                   win_end   = bins[unique(bin_id)] + window_size,
                   count_sum = sum(read_count)), by = bin_id]

    out
  }, by = .(chrom, strand)]
}
```


```{r}
sliding_results <- slide_window_dt(polya_counts, window_size = 1, step = 1)
sliding_results <- sliding_results[count_sum > 5]
head(sliding_results)
plot(sliding_results$count_sum, pch = ".")
write.table(sliding_results, file = piled_up_results, sep = "\t", row.names = F, quote = F)
```
